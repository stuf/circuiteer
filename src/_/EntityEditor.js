import { useSelector, useDispatch } from 'react-redux';
import * as L from 'partial.lenses';
import Toggle from './Toggle';
import { toggleEntity } from 'state/editor';

function EntityEditor(props) {
  const update = useDispatch();

  const currentId = useSelector(L.get(['editor', 'current']));
  const current = useSelector(
    L.get(['editor', 'entities', L.find(o => o.id === currentId)]),
  );

  return (
    <>
      <div className="px-4 py-2">
        <section>
          <header className="font-bold">{currentId}</header>

          <div className="space-y-2">
            <div>
              <Toggle
                checked={current.enabled}
                label="Enabled"
                onChange={() => {
                  update(toggleEntity({ id: currentId }));
                }}
              />
            </div>
          </div>
        </section>
      </div>
    </>
  );
}

export default EntityEditor;
