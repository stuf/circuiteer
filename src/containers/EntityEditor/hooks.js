import * as L from 'partial.lenses';
import { useDispatch, useSelector } from 'react-redux';
import { toggleEntity, deleteEntity } from 'state/editor';
import { modules } from 'config';

/**
 * @param {any} state
 * @return {IEntity[]}
 */
export const selectEntities = state => L.get(['editor', 'entities'], state);

/**
 *
 * @param {*} state
 * @returns {(string | undefined)}
 */
export const selectCurrent = state => L.get(['editor', 'current'], state);

export function useCurrentEntity() {
  const update = useDispatch();

  const items = useSelector(selectEntities);
  const _current = useSelector(selectCurrent);
  const current = items.find(x => x.id === _current);

  if (!current?.module) {
    return {
      current: null,
      module: null,
      toggleCurrent: () => {
        throw new Error(
          [
            "Can not toggle a null entity. You shouldn't be seeing this error",
            "unless you're running tests or you are intentionally doing something",
            'strange.',
          ].join(''),
        );
      },
    };
  }

  const module = modules[current.module];

  return {
    current,
    module,
    toggleCurrent: () => update(toggleEntity({ id: _current })),
    deleteCurrent: () => update(deleteEntity({ id: _current })),
  };
}
