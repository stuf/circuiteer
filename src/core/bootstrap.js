import { useEffect } from 'react';
import { useDispatch } from 'react-redux';

import { setObjects } from 'state/objects';
import { setGameObjects } from 'state/game-entities';
import { setLocations } from 'state/location';
import { setFlags } from 'state/options';
import { setTiers } from 'state/tier';

import objects from 'dev/objects';
import { objects as gameObjects } from 'dev/game-entities';
import locationObjects from 'dev/location';
import { flags } from 'dev/options';
import tiers from 'dev/tier';

/**
 * @todo As alternative, just initialize the store with a specific state
 */
export function useApplicationBootstrap() {
  const update = useDispatch();

  useEffect(() => {
    update(setObjects(objects));
    update(setGameObjects(gameObjects));
    update(setLocations(locationObjects));
    update(setFlags(flags));
    update(setTiers(tiers));
  }, [update]);
}
