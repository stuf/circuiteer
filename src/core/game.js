export const Difficulty = {
  EASY: 1,
  MEDIUM: 2,
  HARD: 3,
  VERY_HARD: 4,
};

export const Wind = {
  VERY_LOW: 0.25,
  LOW: 0.5,
  MEDIUM: 1,
  HIGH: 1.5,
  VERY_HIGH: 1.75,
};

export const Sun = {
  VERY_LOW: 0.25,
  LOW: 0.5,
  MEDIUM: 1,
  HIGH: 1.5,
  VERY_HIGH: 1.75,
};
