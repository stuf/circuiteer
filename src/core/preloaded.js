export const preloadedState = {
  application: {
    splash: {
      delay: 1000,
      show: true,
    },
  },
  drag: {
    dragging: false,
    pos: [0, 0],
    size: [4, 4],
  },
  editor: {
    dragging: false,
    current: null,
    entities: [
      {
        id: '262ec14e-940b-41b3-8c4a-6958afff5332',
        pos: [2, 2],
        module: 'shelter',
        enabled: false,
      },
      {
        id: '9d2d6e06-7e1f-4021-b4cb-dbbb7b6515d0',
        pos: [9, 1],
        module: 'mediumWindTurbine',
        enabled: true,
      },
      {
        id: '2fb98a1d-0738-4def-bf52-f0c91a0e1e45',
        pos: [1, 9],
        module: 'mediumSolarPanel',
        enabled: false,
      },
      {
        id: 'ff6de281-5345-4ff4-af7c-eb9f2afd8694',
        module: 'mediumShredder',
        pos: [9, 5],
        enabled: true,
      },
      {
        id: '5b4c6fc0-5066-4962-973b-9d62316e4fa7',
        module: 'mediumShredder',
        pos: [5, 9],
        enabled: false,
      },
      {
        pos: [22, 2],
        module: 'qtRtg',
        enabled: true,
        id: '96da73e5-9776-44d6-b526-6ed7f47ac9ee',
      },
      {
        pos: [22, 4],
        module: 'qtRtg',
        enabled: true,
        id: '6fc5d160-d017-4db5-8279-63725936ee23',
      },
      {
        pos: [22, 6],
        module: 'rtg',
        enabled: true,
        id: '45fa3e7e-9e1e-4188-ac17-809a3a5847b8',
      },
      {
        pos: [22, 9],
        module: 'rtg',
        enabled: true,
        id: '4d7f50d9-1fb7-4fe7-bc30-001fb94eed41',
      },
      {
        pos: [12, 2],
        module: 'xlargeWindTurbine',
        enabled: true,
        id: '641f9330-d802-45a5-80f9-0b7a45a59a10',
      },
      {
        pos: [11, 9],
        module: 'mediumGenerator',
        enabled: true,
        id: '258a9d97-ac01-4849-84a1-381a2c915fc1',
      },
      {
        pos: [15, 9],
        module: 'mediumGenerator',
        enabled: true,
        id: 'c03c7c14-416e-4ec8-b6b9-29b656573c98',
      },
      {
        pos: [19, 9],
        module: 'mediumBattery',
        enabled: true,
        id: '58a2e641-e2f8-4fba-a079-ff937e760837',
      },
      {
        pos: [19, 5],
        module: 'mediumSolarPanel',
        enabled: true,
        id: 'db75958a-c48b-4a0b-b91e-2c8b6381c1c4',
      },
      {
        pos: [36, 3],
        module: 'largeSolarPanel',
        enabled: true,
        id: '83fb60a4-c57a-4d79-8f99-0d6faf7d3912',
      },
    ],
  },
  grid: {
    size: [48, 48],
  },
  location: {
    locations: [
      {
        id: 'sylva',
        difficulty: 'easy',
        wind: 1,
        sun: 1,
        cycle: [720, 60],
      },
      {
        id: 'desolo',
        difficulty: 'easy',
        wind: 0.5,
        sun: 1.5,
        cycle: 115,
      },
      {
        id: 'calidor',
        difficulty: 'medium',
        wind: 0.5,
        sun: 1.75,
      },
      {
        id: 'vesania',
        difficulty: 'medium',
        wind: 0.5,
        sun: 1.5,
      },
      {
        id: 'novus',
        difficulty: 'medium',
        wind: 1.5,
        sun: 1.5,
      },
      {
        id: 'glacio',
        difficulty: 'hard',
        wind: 1.75,
        sun: 0.25,
      },
      {
        id: 'atrox',
        difficulty: 'veryHard',
        wind: 0.5,
        sun: 0.5,
      },
    ],
    current: 'calidor',
  },
  module: {
    modules: {
      smallWindTurbine: {
        id: 'smallWindTurbine',
        shortId: 'swt',
        power: 1.5,
        powerType: 'wind',
        tier: 1,
        size: [1, 1],
      },
      smallSolarPanel: {
        id: 'smallSolarPanel',
        shortId: 'ss',
        power: 1,
        powerType: 'sun',
        tier: 1,
        size: [1, 1],
      },
      smallGenerator: {
        id: 'smallGenerator',
        shortId: 'sg',
        power: 2,
        powerType: 'powered',
        tier: 1,
        size: [1, 1],
      },
      qtRtg: {
        id: 'qtRtg',
        shortId: 'qr',
        power: 1,
        powerType: 'always',
        tier: 1,
        size: [1, 1],
      },
      smallBattery: {
        id: 'smallBattery',
        shortId: 'sb',
        power: 1,
        powerType: 'powered',
        capacity: 32,
        tier: 1,
        size: [1, 1],
      },
      portableOxygenator: {
        id: 'portableOxygenator',
        shortId: 'po',
        power: -1,
        tier: 1,
        size: [1, 1],
      },
      fieldShelter: {
        id: 'fieldShelter',
        shortId: 'fs',
        power: 0.5,
        powerType: 'always',
        tier: 2,
        size: [2, 2],
      },
      mediumWindTurbine: {
        id: 'mediumWindTurbine',
        shortId: 'mwt',
        power: 5,
        powerType: 'wind',
        tier: 2,
        size: [2, 2],
      },
      mediumSolarPanel: {
        id: 'mediumSolarPanel',
        shortId: 'msp',
        power: 4,
        powerType: 'sun',
        tier: 2,
        size: [2, 2],
      },
      mediumGenerator: {
        id: 'mediumGenerator',
        shortId: 'mg',
        power: 9,
        powerType: 'powered',
        tier: 2,
        size: [2, 2],
      },
      mediumBattery: {
        id: 'mediumBattery',
        shortId: 'mb',
        power: 5,
        powerType: 'powered',
        capacity: 512,
        tier: 2,
        size: [2, 2],
      },
      rtg: {
        id: 'rtg',
        shortId: 'rtg',
        power: 4,
        powerType: 'always',
        tier: 2,
        size: [2, 2],
      },
      autoArm: {
        id: 'autoArm',
        shortId: 'aa',
        power: -1,
        tier: 2,
        size: [2, 2],
      },
      mediumShredder: {
        id: 'mediumShredder',
        shortId: 'ms',
        power: -5,
        tier: 2,
        size: [2, 2],
      },
      largeWindTurbine: {
        id: 'largeWindTurbine',
        shortId: 'lwt',
        power: 10,
        powerType: 'wind',
        tier: 3,
        size: [4, 4],
      },
      largeSolarPanel: {
        id: 'largeSolarPanel',
        shortId: 'lsp',
        power: 8,
        powerType: 'sun',
        tier: 3,
        size: [4, 4],
      },
      researchChamber: {
        id: 'researchChamber',
        shortId: 'rc',
        power: -2,
        tier: 3,
        size: [4, 4],
      },
      smeltingFurnace: {
        id: 'smeltingFurnace',
        shortId: 'sf',
        power: -5,
        tier: 3,
        size: [4, 4],
      },
      soilCentrifuge: {
        id: 'soilCentrifuge',
        shortId: 'sc',
        power: -6,
        tier: 3,
        size: [4, 4],
      },
      chemistryLab: {
        id: 'chemistryLab',
        shortId: 'cl',
        power: -10,
        tier: 3,
        size: [4, 4],
      },
      atmosphericCondenser: {
        id: 'atmosphericCondenser',
        shortId: 'ac',
        power: -20,
        tier: 3,
        size: [4, 4],
      },
      largeShredder: {
        id: 'largeShredder',
        shortId: 'ls',
        power: -7.5,
        tier: 3,
        size: [4, 4],
      },
      shelter: {
        id: 'shelter',
        shortId: 's',
        power: 1,
        powerType: 'always',
        size: [8, 8],
        tier: 4,
      },
      xlargeWindTurbine: {
        id: 'xlargeWindTurbine',
        shortId: 'xlwt',
        power: 17,
        powerType: 'wind',
        tier: 4,
        size: [8, 8],
      },
      solarArray: {
        id: 'solarArray',
        shortId: 'sa',
        power: 14,
        tier: 4,
        size: [8, 8],
      },
      extraLargeShredder: {
        id: 'extraLargeShredder',
        shortId: 'xls',
        power: -10,
        tier: 4,
        size: [8, 8],
      },
      wreckedSolarArray: {
        id: 'wreckedSolarArray',
        shortId: 'wsa',
        power: 64,
        powerType: 'sun',
        tier: -1,
        size: [16, 16],
      },
    },
  },
  options: {
    flags: {
      hideInvalid: false,
      showEditor: true,
      showEfficiencyAsMultiplier: true,
      showPowerStatus: true,
      showShoppingList: true,
    },
  },
  experiment: {
    ids: [],
    entities: {},
  },
};
