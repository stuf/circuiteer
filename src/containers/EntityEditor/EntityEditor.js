import { useState } from 'react';
import { useTranslation } from 'react-i18next';

import Toggle from '_/Toggle';
import { useCurrentEntity } from './hooks';

export function EntityEditor(props) {
  const { current, toggle } = useCurrentEntity();
  const [state, setState] = useState({ showDebug: false });
  const { t } = useTranslation();

  return (
    <aside className="absolute top-2 right-2 bg-white shadow z-10 text-xs w-96">
      <header className="relative px-4 py-2 text-base font-bold">
        {t('common:selected')}
        <Toggle
          className="absolute flex items-center -inset-y-0 right-4 text-sm"
          checked={current.enabled}
          label={t('common:enable')}
          onChange={toggle}
        />
      </header>

      <div className="px-4 py-2">
        <dl
          className="gap-2"
          style={{ display: 'grid', gridTemplateColumns: '1fr 4fr' }}
        >
          <dt>ID</dt>
          <dd className="font-mono">{current.id}</dd>

          <dt>Position</dt>
          <dd className="font-mono">
            (x={current.pos[0]}, y={current.pos[1]})
          </dd>

          <dt>Module</dt>
          <dd>
            <dl style={{ display: 'grid', gridTemplateColumns: '1fr 4fr' }}>
              <dt>Name</dt>
              <dd>{current.module.name}</dd>

              <dt>Power</dt>
              <dd>{current.module.power} U/s</dd>

              <dt>Type</dt>
              <dd>{current.module.powerType}</dd>

              <dt>Tier</dt>
              <dd>{current.module.tier}</dd>
            </dl>
          </dd>
        </dl>
      </div>

      <div className="px-4 py-2">
        <fieldset>
          <legend>
            <Toggle
              label="Show debug"
              checked={state.showDebug}
              onChange={() =>
                setState(s => ({ ...s, showDebug: !s.showDebug }))
              }
            />
          </legend>

          {state.showDebug && (
            <pre className="mt-4">{JSON.stringify(current, null, 2)}</pre>
          )}
        </fieldset>
      </div>
    </aside>
  );
}
