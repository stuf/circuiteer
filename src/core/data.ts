import { BaseMaterial, IBaseMaterial, MaterialType } from 'data/game/material';

const naturals: Record<string, IBaseMaterial> = (type => ({
  soil: { id: 'soil', type },
  organic: { id: 'organic', type },
  compound: { id: 'compound', type },
  resin: { id: 'resin', type },
  clay: { id: 'clay', type },
  quartz: { id: 'quartz', type },
  ammonium: { id: 'ammonium', type },
  graphite: { id: 'graphite', type },
  laterite: { id: 'laterite', type },
  astronium: { id: 'astronium', type },

  // Location-specific natural resources
  sphalerite: { id: 'sphalerite', type },
  wolframite: { id: 'wolframite', type },
  malachite: { id: 'malachite', type },
  lithium: { id: 'lithium', type },
  hematite: { id: 'hematite', type },
  titanite: { id: 'titanite', type },
}))(MaterialType.NATURAL);

export type NaturalKeys = keyof typeof naturals;

export type NaturalResources = Record<NaturalKeys, IBaseMaterial>;

export const NaturalResource: NaturalResources = Object.entries(
  naturals,
).reduce(
  (o, [k, mat]) => ({ ...o, [k]: new BaseMaterial(mat) }),
  {} as NaturalResources,
);

//
