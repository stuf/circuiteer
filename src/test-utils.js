import { render as renderRtl } from '@testing-library/react';
import { format } from 'util';
import { Provider } from 'react-redux';
import { preloadedState } from './core/preloaded';
import * as reducers from './state';
import { configureStore } from '@reduxjs/toolkit';

import './test/index';

//

export function render(vdom, { initialState, store, ...renderOptions } = {}) {
  const rootStore =
    store ||
    configureStore({
      preloadedState: initialState || preloadedState,
      reducer: reducers,
    });

  function Wrapper({ children }) {
    return <Provider store={rootStore}>{children}</Provider>;
  }

  return renderRtl(vdom, { wrapper: Wrapper, ...renderOptions });
}

//

export function suppressPropTypeWarnings() {
  const err = console.error.bind(console.error);

  beforeAll(() => {
    console.error = (msg, ...rest) => {
      const errorMsg = format(msg, ...rest);
      if (!errorMsg.toString().includes('Warning: Failed prop type')) {
        err(errorMsg);
      }
    };
  });

  afterAll(() => {
    console.error = err;
  });
}
