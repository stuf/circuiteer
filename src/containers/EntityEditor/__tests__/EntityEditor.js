import * as L from 'partial.lenses';
import { render } from 'test-utils';
import { fireEvent, createEvent, logRoles } from '@testing-library/react';
import { EntityEditor } from '../EntityEditor';
import { preloadedState } from 'core/preloaded';
import { act } from 'react-dom/test-utils';

it('should show an error if attempting to mount without having a selected entity', async () => {
  const initialState = L.set(['editor', 'current'], null, preloadedState);

  const { container, findByRole, findByText } = render(<EntityEditor />, {
    initialState,
  });

  expect(await findByRole('banner')).not.toBeUndefined();
  expect(await findByText(/something is wrong/i)).not.toBeUndefined();

  expect(container).toMatchSnapshot();
});

it('should show the editor with the current selected', async () => {
  const id = '262ec14e-940b-41b3-8c4a-6958afff5332';
  const initialState = L.set(['editor', 'current'], id, preloadedState);

  const { container, findByRole } = render(<EntityEditor />, {
    initialState,
  });

  const _id = await findByRole('definition', { name: 'ID' });
  expect(_id.textContent).toBe(id);

  expect(container).toMatchSnapshot();

  const _debug = await findByRole('switch', { name: /show debug/i });

  act(() => {
    fireEvent(_debug, createEvent.click(_debug));
  });

  await findByRole('switch', {
    name: /show debug/i,
    checked: true,
  });
});
