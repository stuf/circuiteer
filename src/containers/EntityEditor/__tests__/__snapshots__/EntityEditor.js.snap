// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show an error if attempting to mount without having a selected entity 1`] = `
<div>
  <aside
    class="absolute top-2 right-2 bg-white shadow-md border-2 rounded-md z-10 text-xs w-96"
  >
    <header>
      Something is wrong
    </header>
    <div>
      You shouldn't be seeing this, as the entity editor should only be shown when having selected an entity.
    </div>
  </aside>
</div>
`;

exports[`should show the editor with the current selected 1`] = `
<div>
  <aside
    class="absolute top-2 right-2 bg-white shadow-md border-2 rounded-md z-10 text-xs w-96"
  >
    <header
      class="relative px-4 py-2 text-base font-bold"
    >
      game:module.shelter
      <div
        class="toggle--wrap"
      >
        <div
          class="toggle toggle--root absolute flex items-center -inset-y-0 right-4 text-sm group"
        >
          <button
            aria-checked="false"
            class="
        toggle__switch
        bg-gray-200
      "
            id="headlessui-switch-1"
            role="switch"
            tabindex="0"
          >
            <span
              class="sr-only"
            >
              common:enable
            </span>
            <span
              class="toggle__tick"
            />
          </button>
          <span
            class="toggle__label"
          >
            common:enable
          </span>
        </div>
      </div>
    </header>
    <div
      class="px-4 py-2"
    >
      <dl
        class="gap-2"
        style="display: grid; grid-template-columns: 1fr 4fr;"
      >
        <dt
          id="entity-id"
        >
          ID
        </dt>
        <dd
          aria-labelledby="entity-id"
          class="font-mono"
        >
          262ec14e-940b-41b3-8c4a-6958afff5332
        </dd>
        <dt
          id="entity-position"
        >
          Position
        </dt>
        <dd
          aria-labelledby="entity-position"
          class="font-mono"
        >
          (x=
          2
          , y=
          2
          )
        </dd>
        <dt
          id="entity-size"
        >
          Size
        </dt>
        <dd
          aria-labelledby="entity-size"
          class="font-mono"
        >
          (w=
          8
          , h=
          8
          )
        </dd>
        <dt
          id="entity-module"
        >
          Module
        </dt>
        <dd
          aria-labelledby="entity-module"
        >
          <dl
            style="display: grid; grid-template-columns: 1.4fr 4fr;"
          >
            <dt
              id="module-id"
            >
              Module ID
            </dt>
            <dd
              aria-labelledby="module-id"
              class="font-mono"
            >
              shelter
            </dd>
            <dt
              id="module-power"
            >
              Power
            </dt>
            <dd
              aria-labelledby="module-power"
              class="font-mono"
            >
              1
               U/s
            </dd>
            <dt
              id="module-type"
            >
              Type
            </dt>
            <dd
              aria-labelledby="module-type"
              class="font-mono"
            >
              always
            </dd>
            <dt
              id="module-tier"
            >
              Tier
            </dt>
            <dd
              aria-labelledby="module-tier"
              class="font-mono"
            >
              4
            </dd>
          </dl>
        </dd>
      </dl>
    </div>
    <div
      class="px-4 py-2"
    >
      <button
        aria-label="Delete"
        class="button button--with-icon"
        type="button"
      >
        <span
          aria-hidden="true"
          class="__icon-wrapper"
        >
          <span
            class="icon icon--uiflex items-center justify-center"
            style="width: 20px; height: 20px;"
          >
            <span
              class="material-icons material-icons-outlined"
              style="font-size: 20px;"
            >
              delete
            </span>
          </span>
        </span>
        Delete
      </button>
    </div>
    <div
      class="px-4 py-2"
    >
      <fieldset>
        <legend>
          <div
            class="toggle--wrap"
          >
            <div
              class="toggle toggle--root group"
            >
              <button
                aria-checked="false"
                class="
        toggle__switch
        bg-gray-200
      "
                id="headlessui-switch-2"
                role="switch"
                tabindex="0"
              >
                <span
                  class="sr-only"
                >
                  Show debug
                </span>
                <span
                  class="toggle__tick"
                />
              </button>
              <span
                class="toggle__label"
              >
                Show debug
              </span>
            </div>
          </div>
        </legend>
      </fieldset>
    </div>
  </aside>
</div>
`;
